internal/host/plugin/repository_host_set.go

```diff
@@ -83,49 +125,172 @@ func (r *Repository) CreateSet(ctx context.Context, scopeId string, s *HostSet,
+		return nil, errors.New(ctx, errors.InvalidParameter, op, "searching for both a host set id and a catalog id is not supported")
 	}
-	var sets []*HostSet
-	err := r.reader.SearchWhere(ctx, &sets, "catalog_id = ?", []interface{}{catalogId}, db.WithLimit(limit))
+
+	opts, err := host.GetOpts(opt...)
 	if err != nil {
 		return nil, errors.Wrap(ctx, err, op)
 	}
+
+	limit := r.defaultLimit
+	if opts.WithLimit != 0 {
+		// non-zero signals an override of the default limit for the repo.
+		limit = opts.WithLimit
+	}
+
+	args := make([]interface{}, 0, 1)
+	var where string
+
+	switch {
+	case publicId != "":
+		where, args = "public_id = ?", append(args, publicId)
+	default:
+		where, args = "catalog_id = ?", append(args, catalogId)
+	}
+
+	dbArgs := []db.Option{db.WithLimit(limit)}
+
+	if opts.WithOrderByCreateTime {
+		if opts.Ascending {
+			dbArgs = append(dbArgs, db.WithOrder("create_time asc"))
+		} else {
+			dbArgs = append(dbArgs, db.WithOrder("create_time"))
+		}
+	}
+
+	var aggHostSets []*hostSetAgg
+	if err := r.reader.SearchWhere(ctx, &aggHostSets, where, args, dbArgs...); err != nil {
+		return nil, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("in %s", publicId)))
+	}
+
+	if len(aggHostSets) == 0 {
+		return nil, nil
+	}
+
+	sets := make([]*HostSet, 0, len(aggHostSets))
+	for _, agg := range aggHostSets {
+		hs := allocHostSet()
+		hs.PublicId = agg.PublicId
+		hs.CatalogId = agg.CatalogId
+		hs.Name = agg.Name
+		hs.Description = agg.Description
+		hs.CreateTime = agg.CreateTime
+		hs.UpdateTime = agg.UpdateTime
+		hs.Version = agg.Version
+		hs.Attributes = agg.Attributes
+		if agg.PreferredEndpoints != "" {
+			eps := strings.Split(agg.PreferredEndpoints, aggregateDelimiter)
+			if len(eps) > 0 {
+				// We want to protect against someone messing with the DB
+				// and not panic, so we do a bit of a dance here
+				var sortErr error
+				sort.Slice(eps, func(i, j int) bool {
+					epi := strings.Split(eps[i], priorityDelimiter)
+					if len(epi) != 2 {
+						sortErr = errors.New(ctx, errors.NotSpecificIntegrity, op, fmt.Sprintf("preferred endpoint %s had unexpected fields", eps[i]))
+						return false
+					}
+					epj := strings.Split(eps[j], priorityDelimiter)
+					if len(epj) != 2 {
+						sortErr = errors.New(ctx, errors.NotSpecificIntegrity, op, fmt.Sprintf("preferred endpoint %s had unexpected fields", eps[j]))
+						return false
+					}
+					indexi, err := strconv.Atoi(epi[0])
+					if err != nil {
+						sortErr = errors.Wrap(ctx, err, op)
+						return false
+					}
+					indexj, err := strconv.Atoi(epj[0])
+					if err != nil {
+						sortErr = errors.Wrap(ctx, err, op)
+						return false
+					}
+					return indexi < indexj
+				})
+				if sortErr != nil {
+					return nil, sortErr
+				}
+				for i, ep := range eps {
+					// At this point they're in the correct order, but we still
+					// have to strip off the priority
+					eps[i] = strings.Split(ep, priorityDelimiter)[0]
```

Comment:
```suggestion
					eps[i] = strings.Split(ep, priorityDelimiter)[1]
```

Reply:
Ugh I think I fell victim to the nest test framework. I definitely ran tests and they definitely worked.